src/utils/todoUtils.ts

```diff
@@ -0,0 +1,239 @@
+import { Todo, Subtask, Checklist } from '@/types/todo';
+
+export function calculateSubtaskProgress(todo: Todo): number {
+  if (todo.subtasks.length === 0) return 0;
+  const completed = todo.subtasks.filter(s => s.completed).length;
+  return Math.round((completed / todo.subtasks.length) * 100);
+}
+
+export function calculateSubtaskProgress2(todo: Todo): number {
```

Comment:
[HIGH] Multiple functions (`calculateSubtaskProgress`, `calculateChecklistProgress`) are duplicated with similar logic.

Suggested change:

```suggestion
export function calculateSubtaskProgress(todo: Todo): number {
    if (todo.subtasks.length === 0) return 0;
    const completed = todo.subtasks.filter(s => s.completed).length;
    return Math.round((completed / todo.subtasks.length) * 100);
}
```